lib/common/issues/src/broker.rs

```diff
@@ -0,0 +1,132 @@
+use std::sync::Arc;
+
+use tokio::task;
+
+use crate::typemap::TypeMap;
+
+pub trait Subscriber<E> {
+    fn notify(&self, event: Arc<E>);
+}
+
+struct SubscriberMap(TypeMap);
+
+type DynSubscriber<E> = Box<dyn Subscriber<E> + Send + Sync>;
+type SubscriVec<E> = Vec<Arc<DynSubscriber<E>>>;
+
+impl SubscriberMap {
+    fn new() -> Self {
+        Self(TypeMap::new())
+    }
+    fn push<E: 'static>(&mut self, subscriber: DynSubscriber<E>) {
+        if !self.0.has::<SubscriVec<E>>() {
+            self.0.insert(SubscriVec::<E>::new());
+        }
+        let sub = Arc::new(subscriber);
+
+        self.0.get_mut::<SubscriVec<E>>().unwrap().push(sub);
+    }
+
+    fn get<E: 'static>(&self) -> Option<&SubscriVec<E>> {
+        self.0.get()
+    }
+}
+
+pub struct EventBroker {
+    subscriptions: SubscriberMap,
+}
+
+impl EventBroker {
+    pub fn new() -> Self {
+        Self {
+            subscriptions: SubscriberMap::new(),
+        }
+    }
+    pub fn add_subscriber<E: 'static>(&mut self, subscriber: DynSubscriber<E>) {
+        self.subscriptions.push(subscriber);
+    }
+
+    /// Notify all subscribers of the event. Spawns one tokio task per subscriber, so that it can be completed in the background.
+    pub fn publish_async<E: Send + Sync + 'static>(&self, event: E) {
```

Comment:
will you remove `publish_async` after our conversation?

it seems all `notify` can run async. computation in the background.

Reply:
Yeah I think we can remove it for now. Ideally I'd want this kind of `publish_async` to return a handle so that we can choose whether we want to await for all the notifications to be handled or not. But that sounds like a refactor/discussion we can have later.

We are not using this function anywhere right now :)